---
title: "Gowinda outputs"
author: "Harrison Ostridge"
date: "23/07/2021"
output: html_document
---

This script analyses gene set enrichment results from running Gowinda. I highlight cases of nominal (FDR<=0.1) and significant (FDR<=0.5) enrichment.

## Figure explanation

Results from each gene set database tested are given under subtitles. I also write a little about interesting patterns I identify under the graphs with notes on the functions of gene sets and the signal detected. 

The bar charts show the FDR for each gene set, the colour of the bars also corresponds to the FDR so shorter bars (i.e. smaller FDRs) are more red just to aid identification of significant results when looking at lots of graphs together. On the right of the graph I also give the expected and observed number of hits to assess the effect size.

Each row corresponds to a different subspecies dataset (all='all samples', ce='central and eastern', n='Nigeria-Cameroon', w='western') and each column corresponds to a different empirical p value threshold (0.5%, 0.1%, 0.05%).

I also plot Venn diagrams to get an understanding of whether enrichment of multiple gene sets are independent signals or not.

### Setup

```{r setup}
rm(list=ls())
knitr::opts_knit$set(root.dir = '~/OneDrive - University College London/Projects/Ostridge_PanAf/gowinda/baypass_core') 
```

### Library

```{r library, message=FALSE, warning=FALSE}
library(data.table)
options(datatable.fread.datatable=FALSE)
library(dplyr)
library(ggplot2)
library(wesanderson)
library(VennDiagram)
library(UpSetR)
library(tidyverse)
library(ggh4x)
source("scripts/plot_gowinda.R")

gowinda_out_dir="output/gowinda_output/"
```

```{r}
stat='fpr'
subsps=c('all', 'ce', 'n', 'w')

if(stat=='p'){
  tails=c('0.5pct-cov_cor', '0.1pct-cov_cor', '0.05pct-cov_cor')
  stat_name="Empirical tails"
}else if(stat=='fpr'){
  tails=c('fpr0.5pct.non-genic_1000bp.flanks', 'fpr0.1pct.non-genic_1000bp.flanks', 'fpr0.05pct.non-genic_1000bp.flanks')
  stat_name="Candidate FPR tails"
}else{
  stop("Define stat as either 'p' (empircical p value) or fpr (FPR)")
}
```

### Define Function

```{r}
plot_gowinda_all=function(subsps, tails, gene.set){
  for(subsp in subsps){
    cat("-", subsp, "\n")
    for(tail in tails){
      gowinda_output_file=paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.", gene.set)
      if(file.exists(gowinda_output_file) & file.size(gowinda_output_file) > 0){
        df=fread(gowinda_output_file)
        if(nrow(df>0)){plot_gowinda(df, title=paste0(subsp, ": ", "\n", tail, " candidates"))}
      }
    }
  }
}
```

# Gene Function Databases

## GO

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='go')
```

### Central and Eastern

#### Transmission_across_Electrical_Synapses and Electric_Transmission_Across_Gap_Junctions

Signal
  
- Nominal enrichment at 0.1% (FDR=0.11)

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
subsp='ce'
tail='fpr0.1pct.non-genic_1000bp.flanks'
gene.set='go'
gowinda_out=fread(paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.", gene.set), sep='\t', header=F)
genes=gowinda_out[1:3,10]
genes=strsplit(genes, ",")
names(genes)=gowinda_out[1:3,9]
cat(subsp, gene.set, tail)
upset(fromList(genes), sets = rev(names(genes)), keep.order = TRUE, order.by = "degree")
```

## Reactome

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='reac')
```

### Central and Eastern

#### Unfolded protein response

Signal
  
- Nominal enrichment at 0.5% (FDR=0.08)

## GWAS

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='gwas')
```

## KEGG

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='kegg')
```

## Tissue expression data

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='expr')
```

### Western

#### Spleen

Signal
  
- Significant enrichment at 0.1% and 0.05%

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
subsp='w'
tail='fpr0.1pct.non-genic_1000bp.flanks'
gene.set='expr'
gowinda_out=fread(paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.", gene.set), sep='\t', header=F)
genes=gowinda_out[1:11,10]
genes=strsplit(genes, ",")
names(genes)=gowinda_out[1:11,9]
cat(subsp, gene.set, tail)
upset(fromList(genes), sets = rev(names(genes)), keep.order = TRUE, order.by = "degree")
```

## Tissue expression data - semi-unique

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='expr.5overlap')
```

## Phenotype database

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='phen')
```

### All subspecies

#### Lung related categories

##### Signal

- Significant enrichment at 0.1% tail

- Large overlpa between categories - driven by 6 genes in total

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
subsp='all'
tail='fpr0.1pct.non-genic_1000bp.flanks'
gene.set='phen'
gowinda_out=fread(paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.", gene.set), sep='\t', header=F)
genes=gowinda_out[1:8,10]
genes=strsplit(genes, ",")
names(genes)=gowinda_out[1:8,9]
cat(subsp, gene.set, tail)
upset(fromList(genes), sets = rev(names(genes)), keep.order = TRUE, order.by = "degree")
```

# Pathogen Datasets 

We do not really see any strong signals when running hypothesis free tests using large gene function databases. Our previous work has also pathogens to be important section pressures in chimps at different time scales and so we decided to run some more hypothesis driven tests, investigating enrichment of genes we know are important in pathogen infection. 

## Viruses

### VIPs

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='vips')
```

Function 

- Manually curated and highthroughoput list of genes which code for proteins which physically bind with viruses.

Signal 

- No significant enrichment.

## Hypothesis Driven Pathogens (including ebel 2017)

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='pathogen_ebel2017')
```

### Central-eastern

#### HSV-1

##### Signal

- Enriched at 0.5% tails (FDR=0.05)

### Western

#### COVID genes

##### Signal

- Enriched at 0.05% tail (FDR=0.05)

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
subsp='w'
tail='fpr0.05pct.non-genic_1000bp.flanks'
gene.set='pathogen'
gowinda_out=fread(paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.", gene.set), sep='\t', header=F)
genes=gowinda_out[1:3,10]
genes=strsplit(genes, ",")
names(genes)=gowinda_out[1:3,9]
cat(subsp, gene.set, tail)
upset(fromList(genes), sets = rev(names(genes)), keep.order = TRUE, order.by = "degree")
```

- NB: a potential confounding factor is that gene aliases may reduce the amount of overlap reported in the above figure but I would expect this effect to be rather small.

## Immunity genes

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='imm')
```

## Dehydration genes

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
plot_gowinda_all(subsps=subsps, tails=tails, gene.set='dehy')
```

# Heat plot

```{r}
datasets=c("go", "kegg", "reac", "gwas", "expr", "phen", "vips", "pathogen_ebel2017", "imm", "dehy")
```

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
total_gowinda=NULL
p_adjust=F

for(subsp in subsps){
  for(tail in tails){
    tail_name=NULL
    if(grepl("0\\.5pct", tail)){
      tail_name="0.5%"
    }else if(grepl("0\\.1pct", tail)){
      tail_name="0.1%"
    }else if(grepl("0\\.05pct", tail)){
        tail_name="0.05%"
    }
    for(dataset in datasets){
      if(dataset %in% c("go", "kegg", "reac", "gwas", "expr", "phen")){
        category="General"
      }else if(dataset %in% c("vips", "pathogen_ebel2017", "imm")){
        category="Pathogens"
      }else if(dataset %in% c("dehy")){
          category="Dehydration"
      }else{
        category="Other"
      }
      df=data.frame()
      gowinda_output_file=paste0(gowinda_out_dir, "/",subsp,"/",subsp,".",tail,".gowinda_out.",dataset)
      
      if(file.exists(gowinda_output_file) & file.size(gowinda_output_file) > 0){
        df=fread(gowinda_output_file, sep="\t", header=FALSE)
        colnames(df)=c("gene_set", "expected_hits", "observed_hits", "p", "fdr", "No._candidate_genes_in_gene_set", 
                       "No._genes_in_gene_set_and_annotation_and_snp", "No._genes_in_gene_set", "gene_set_description", "candidate_genes")
        df=df[!(df$gene_set %in% c("all", "all_genes", "all.genes") | df$gene_set_description %in% c("all", "all_genes", "all.genes")),]
      }
      if(nrow(df)>0){
        n=nrow(df)
        rows=cbind(data.frame("Subspecies"=rep(subsp, n), "Tail"=rep(tail_name, n),
                              "Category"=rep(category, n), "Dataset"=rep(dataset, n)),
                   df)
        if(is.null(total_gowinda)){
          total_gowinda=rows
        }else{
          total_gowinda=rbind(total_gowinda, rows)
        }
      }
    }
  }
}
total_gowinda$Tail=factor(total_gowinda$Tail, levels=c("0.5%", "0.1%", "0.05%"))
#total_gowinda$Dataset=factor(total_gowinda$Dataset, levels=datasets)
total_gowinda$Category=factor(total_gowinda$Category, levels=c("General", "Pathogens", "Dehydration", "Other"))

#total_gowinda=total_gowinda[total_gowinda$Category!="Other" & total_gowinda$Dataset!="go",]
#total_gowinda=total_gowinda[total_gowinda$Category!="Other",]

# Capitalise first letter
firstup=function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}
total_gowinda$gene_set_description=firstup(total_gowinda$gene_set_description)

if(p_adjust){
  p_adjust_list=list()
  for(subsp in unique(total_gowinda$Subspecies)){
    for(tail in unique(total_gowinda$Tail)){
      tmp_df=total_gowinda[total_gowinda$Subspecies==subsp & total_gowinda$Tail==tail,]
      tmp_df$fdr=p.adjust(tmp_df$p, method = "BH")
      p_adjust_list=append(p_adjust_list, list(tmp_df))
    }
  }
  total_gowinda=do.call(rbind, p_adjust_list)
}

```

#```{r echo=FALSE, out.width="100%", fig.height=3.5, fig.width=5.5}
```{r echo=FALSE, out.width="100%", fig.height=5, fig.width=6}
thresh=0.1
min_n_genes=0
#always_plot_dataset=c("pathogen_ebel2017", "imm")
always_plot_dataset=NULL
do_not_plot=c("dehy")
#cell_fill='observed_hits'
cell_fill='sig'

total_gowinda_comp=merge(complete(total_gowinda[total_gowinda$No._candidate_genes_in_gene_set >= min_n_genes,
                                                c('Subspecies', 'Tail', 'gene_set', 'fdr', 'observed_hits', 
                                                  'No._genes_in_gene_set_and_annotation_and_snp', 'expected_hits')], 
                                  Subspecies, Tail, gene_set), 
                    unique(total_gowinda[,c('Dataset', 'Category', 'gene_set_description', 'gene_set')]), 
                    by='gene_set')

gene_sets=unique(total_gowinda_comp[total_gowinda_comp$fdr<thresh, 'gene_set'])
plot_df=total_gowinda_comp[(total_gowinda_comp$gene_set %in% gene_sets | total_gowinda_comp$Dataset %in% always_plot_dataset) & !(total_gowinda_comp$Dataset %in% do_not_plot),]
#plot_df=plot_df[plot_df$No._candidate_genes_in_gene_set >= min_n_genes,]
plot_df$sig=""
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.05, 'sig']='*'
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.01, 'sig']='**'
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.005, 'sig']='***'
plot_df$observed_hits=paste0(plot_df$observed_hits, plot_df$sig)
## Trim the description so it isn't ridiculously long 
plot_df$gene_set_description=strtrim(plot_df$gene_set_description, 50)
## Set size of text in cells
if(cell_fill=='sig'){
  cell_fill_size=4
  nudge=-0.33
}else{
  cell_fill_size=2.5
  nudge=0
}

#####################################

# Rename
plot_df[plot_df$gene_set_description=="Arrhythmogenic right ventricular cardiomyopathy (ARVC)", 'gene_set_description']="Arrhythmogenic right ventricular cardiomyopathy"
plot_df[plot_df$gene_set_description=="Chronic_obstructive_pulmonary_disease_or_resting_heart_rate_(pleiotropy)", 'gene_set_description']="COPD/resting_heart_rate"
plot_df[plot_df$gene_set_description=="Spherical_equivalent_(joint_analysis_main_effects_and_education_interaction)", 'gene_set_description']="Spherical_equivalent"
plot_df[plot_df$gene_set_description=="Normal_facial_asymmetry_(deformation_magnitude)", 'gene_set_description']="Normal_facial_asymmetry"
plot_df[plot_df$gene_set_description=="SARS-CoV-2: differential_RNA_expression", 'gene_set_description']="SARS-CoV-2: differential expression"
plot_df[plot_df$gene_set_description=="GWAS: Total bilirubin levels in HIV-1 infection", 'gene_set_description']="GWAS: Bilirubin levels in HIV-1 infection"

## Remove references
plot_df$gene_set_description=gsub(" \\(Klunk.*| \\(Rogers.*| \\(Deschamps.*| \\(Daub.*| \\(Ebel.*", 
                                  "", plot_df$gene_set_description)
## Replace "_" with space
plot_df$gene_set_description=gsub("_", " ", plot_df$gene_set_description)
## Order datasets
#plot_df$Dataset=factor(plot_df$Dataset, levels=datasets)
## Order genesets
gene_set_order=c("Airway Obstruction", "Tobacco Use Disorder", "Bronchodilator Agents", "Forced Expiratory Volume", "Aorta",
                 "Immunity genes", "Innate immunity genes")
gene_set_order=rev(gene_set_order)
gene_set_order=gene_set_order[gene_set_order %in% unique(plot_df$gene_set_description)]
other_gene_sets=unique(plot_df$gene_set_description[!(plot_df$gene_set_description %in% gene_set_order)])
plot_df$gene_set_description=factor(plot_df$gene_set_description, levels=c(other_gene_sets[order(other_gene_sets)], gene_set_order))

#####################################

## Rename
plot_df[plot_df$Subspecies=="all", "Subspecies"]="All"
plot_df[plot_df$Subspecies=="ce", "Subspecies"]="C-E"
plot_df[plot_df$Subspecies=="n", "Subspecies"]="N-C"
plot_df[plot_df$Subspecies=="w", "Subspecies"]="W"

plot_df[plot_df$Dataset=="reac", "Dataset"]="Reactome"
plot_df[plot_df$Dataset=="gwas", "Dataset"]="GWAS"
plot_df[plot_df$Dataset=="kegg", "Dataset"]="KEGG"
plot_df[plot_df$Dataset=="pathogen_ebel2017", "Dataset"]="Pathogen-rel."
plot_df[plot_df$Dataset=="expr", "Dataset"]="Tissue expr."
plot_df[plot_df$Dataset=="phen", "Dataset"]="Phenotype"
plot_df[plot_df$Dataset=="imm", "Dataset"]="Immunity"
plot_df[plot_df$Dataset=="dehy", "Dataset"]=""

ggplot(plot_df, aes(x=Tail, y=gene_set_description, fill=fdr))+
      theme_linedraw() +
      theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5), 
            legend.title=element_text(size=10, hjust = 0.5), legend.text=element_text(size=7), 
            axis.title.x = element_text(size=10),
            axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size=7), axis.line = element_blank(), strip.text.y.left = element_text(angle = 0),
            axis.text.y = element_text(size=6),
            strip.background =element_rect(fill="white"), 
            strip.text = element_text(colour = 'black', size=7),
            panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(),
            #legend.position = c(-0.27, 1.14), 
            legend.position = c(-0.27, 1.08), 
            legend.box.background = element_rect(fill="transparent",size=0),legend.key = element_rect(fill = "transparent", colour=NA),
            legend.box = "horizontal", legend.key.height= unit(2.5, 'mm'), legend.key.width= unit(5, 'mm')
            ) +
      labs(title="Genetics-only", x=stat_name, y="") +
      geom_tile() +
      geom_text(aes_string(label=cell_fill), position=position_nudge(y = nudge), size=cell_fill_size) +
      # Add point to indicate significance of 0.1
      geom_point(data=plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.1 & plot_df$fdr>0.05,], size=0.05) +
      scale_fill_gradientn(name = "FDR", colors = rev(wes_palette("Zissou1", type = "continuous")), na.value = "white", limits = c(0,1),
                           guide = guide_colourbar(direction = "horizontal", title.position = "top"), breaks = c(0, 0.5,1)) +
      #facet_grid(Dataset~Subspecies, scales = "free", space='free')
      scale_y_discrete(position = 'right')+ 
      facet_nested(Category+Dataset~Subspecies, scales = "free", space='free', switch = "y")

cat("Min genes per threshold: ", min_n_genes)

#total_gowinda$gene_set_description[grepl("Arrhythmogenic", total_gowinda$gene_set_description)]

# Write supplementary file
write.csv(total_gowinda, "output/gowinda_output/formatted_output/genetics-only_gowinda_results.csv", row.names = FALSE)
```

### dehydration genes

```{r echo=FALSE, out.width="100%", fig.height=1.45, fig.width=5.5}
thresh=0
min_n_genes=0
always_plot_dataset=c("dehy")
do_not_plot=c("")
#cell_fill='observed_hits'
cell_fill='sig'

gene_sets=unique(total_gowinda_comp[total_gowinda_comp$fdr<thresh, 'gene_set'])
plot_df=total_gowinda_comp[(total_gowinda_comp$gene_set %in% gene_sets | total_gowinda_comp$Dataset %in% always_plot_dataset) & !(total_gowinda_comp$Dataset %in% do_not_plot),]
#plot_df=plot_df[plot_df$No._candidate_genes_in_gene_set >= min_n_genes,]
plot_df$sig=""
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.05, 'sig']='*'
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.01, 'sig']='**'
plot_df[!is.na(plot_df$fdr) & plot_df$fdr<0.005, 'sig']='***'
plot_df$observed_hits=paste0(plot_df$observed_hits, plot_df$sig)
## Trim the description so it isn't ridiculously long 
plot_df$gene_set_description=strtrim(plot_df$gene_set_description, 50)
## Set size of text in cells
if(cell_fill=='sig'){
  cell_fill_size=4
  nudge=-0.33
}else{
  cell_fill_size=2.5
  nudge=0
}

#####################################

# Rename
plot_df[plot_df$gene_set_description=="Chronic_obstructive_pulmonary_disease_or_resting_heart_rate_(pleiotropy)", 'gene_set_description']="COPD/resting_heart_rate"
plot_df[plot_df$gene_set_description=="Spherical_equivalent_(joint_analysis_main_effects_and_education_interaction)", 'gene_set_description']="Spherical_equivalent"
plot_df[plot_df$gene_set_description=="Normal_facial_asymmetry_(deformation_magnitude)", 'gene_set_description']="Normal_facial_asymmetry"
plot_df[plot_df$gene_set_description=="SARS-CoV-2: differential_RNA_expression", 'gene_set_description']="SARS-CoV-2: differential expression"
plot_df[plot_df$gene_set_description=="GWAS: Total bilirubin levels in HIV-1 infection", 'gene_set_description']="GWAS: Bilirubin levels in HIV-1 infection"

## Remove references
plot_df$gene_set_description=gsub(" \\(Klunk.*| \\(Rogers.*| \\(Deschamps.*| \\(Daub.*| \\(Ebel.*", 
                                  "", plot_df$gene_set_description)
## Replace "_" with space
plot_df$gene_set_description=gsub("_", " ", plot_df$gene_set_description)
## Order datasets
#plot_df$Dataset=factor(plot_df$Dataset, levels=datasets)
## Order genesets
gene_set_order=c("Airway Obstruction", "Tobacco Use Disorder", "Bronchodilator Agents", "Forced Expiratory Volume", "Aorta",
                 "SARS-CoV-2: differential expression",
                 "Immunity genes", "Innate immunity genes")
gene_set_order=rev(gene_set_order)
gene_set_order=gene_set_order[gene_set_order %in% unique(plot_df$gene_set_description)]
other_gene_sets=unique(plot_df$gene_set_description[!(plot_df$gene_set_description %in% gene_set_order)])
plot_df$gene_set_description=factor(plot_df$gene_set_description, levels=c(other_gene_sets[order(other_gene_sets)], gene_set_order))

#####################################

## Rename
plot_df[plot_df$Subspecies=="all", "Subspecies"]="All"
plot_df[plot_df$Subspecies=="ce", "Subspecies"]="C-E"
plot_df[plot_df$Subspecies=="n", "Subspecies"]="N-C"
plot_df[plot_df$Subspecies=="w", "Subspecies"]="W"
#plot_df$Subspecies=factor(plot_df$Subspecies, levels=c("All", "C-E", "N-C", "W"))

plot_df[plot_df$Dataset=="gwas", "Dataset"]="GWAS"
plot_df[plot_df$Dataset=="kegg", "Dataset"]="KEGG"
plot_df[plot_df$Dataset=="pathogen_ebel2017", "Dataset"]="Pathogen\nrelated"
plot_df[plot_df$Dataset=="expr", "Dataset"]="Tissue expr."
plot_df[plot_df$Dataset=="phen", "Dataset"]="Phenotype"
plot_df[plot_df$Dataset=="imm", "Dataset"]="Immunity"
plot_df[plot_df$Dataset=="dehy", "Dataset"]=""



ggplot(plot_df, aes(x=Tail, y=gene_set_description, fill=fdr))+
      theme_linedraw() +
      theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5), 
            legend.title=element_text(size=10, hjust = 0.5), legend.text=element_text(size=7), 
            axis.title.x = element_text(size=10),
            axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size=7), axis.line = element_blank(), strip.text.y.left = element_text(angle = 0),
            axis.text.y = element_text(size=6),
            strip.background =element_rect(fill="white"), 
            strip.text = element_text(colour = 'black', size=7),
            panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(),
            legend.position = c(-0.23, 2.3), legend.box = "horizontal", legend.key.height= unit(2.5, 'mm'), legend.key.width= unit(5, 'mm')
            ) +
      labs(title="Genetics-only", x=stat_name, y="") +
      geom_tile() +
      geom_text(aes_string(label=cell_fill), position=position_nudge(y = nudge), size=cell_fill_size) +
      scale_fill_gradientn(name = "FDR", colors = rev(wes_palette("Zissou1", type = "continuous")), na.value = "white", limits = c(0,1),
                           guide = guide_colourbar(direction = "horizontal", title.position = "top"), breaks = c(0, 0.5,1)) +
      #facet_grid(Dataset~Subspecies, scales = "free", space='free')
      scale_y_discrete(position = 'right')+ 
      facet_nested(Category+Dataset~Subspecies, scales = "free", space='free', switch = "y")

cat("Min genes per threshold: ", min_n_genes)


#total_gowinda_comp[total_gowinda_comp$Dataset=="dehy",]
```
