#### Setup

```{r setup}
rm(list=ls())
knitr::opts_knit$set(root.dir = '~/OneDrive - University College London/Projects/Ostridge_PanAf/baypass/analysing_baypass_output/baypass_aux/') 
```

#### Parameters

```{r}
subsps=c('all', 'ce', 'w')
env_data='f_over_sum_known_trees'
habitat_col='black'
col_scale=colorRampPalette(c('goldenrod1', 'yellow4', 'darkgreen'))(100)
```

---
title: "`r paste0("BayPass AUX Distributions - ", env_data)`"
subtitle: "`r paste0("Subspecies datasets: ", paste(subsps, collapse = ", "))`"
author: "Harrison Ostridge"
date: "`r Sys.Date()`"
output: html_document
---

Here I plot the distributions of key statistics from running the BayPass AUX model with habitat classifications from Aleman et al. (2020) (using 4 categories) for all subspecies datasets.

Environmental data input files were made with `../../environmental_data/scripts/1.extract_environmental_and_behavioural_data.Rmd` and `../../environmental_data/scripts/2.format_env_file.Rmd`.

BayPass was run on myriad with scripts in `myriad:analysis/phase1and2_exome_analysis/baypass/scripts/mapped.on.target/`. BayPass was run three times with different seeds.

BayPass output files we formatted with `./scripts/format_baypass_aux_output.ipynb` where SNP coordinates, gene annotations and results from multiple independent runs were added. Statistics with no suffix represent results from the focal run (e.g. 'M_Beta'), and statistics from the the two repeat runs have the suffixes '.r1' and '.r2' (e.g. 'M_Beta.r1' and 'M_Beta.r2'). Median values have also been calculated. I also add the results from random runs used to generate nulls.

#### Library

```{r library, message=FALSE, warning=FALSE}
library(data.table)
options(datatable.fread.datatable=FALSE)
library(dplyr)
library(ggplot2)
library(VennDiagram)
library(psych)
library(tidyverse)
library(cowplot)
library(raster)

source("scripts/baypass_aux_tools.R")
source("../scripts/baypass_tools.R")
source("../baypass_core/scripts/candidate_allele_frequency_patterns.R")

env_dir="../../../environmental_data/output/"
formatted_aux_out_dir="output/formatted_baypass_aux_output/"
formatted_core_out_dir="../baypass_core/output/formatted_baypass_core_output/"
allele_freq_dir="../../../allele_frequencies/exome/output/"
annotation_file="../../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs.gtf"
```

#### Read in data

```{r echo=FALSE}
## Environmental data file
env_file=read.csv(paste0("/meta_data.env.pops.imputed.csv"))
### Make a version with a single row per population for plotting (average latitudes and longitudes across communities in a population are used)
env_file.pops=unique(env_file[,c("Population", "Latitude", "Longitude")])
env_file.pops=aggregate(env_file.pops[, c("Latitude", "Longitude")], list(env_file.pops$Population), mean)
colnames(env_file.pops)=c("Population", "Latitude", "Longitude")

# Subspecies specific files
betai.ann=list()
betai=list()
Pdelta=list() #
cov.in=list()
subsp.col=list()
coverage=list()

for(subsp in subsps){
  cat("-", subsp, "\n")
  if(subsp=='all'){subsp.file=''}else{subsp.file=paste0(".", subsp)}
  if(subsp=='n'){miss.pop=0}else{miss.pop=0.3}
  ### Subspecies colours for populations
  pops=unique(fread(paste0(formatted_core_out_dir, "/f5", subsp.file,
          ".pops_missing.pops.", miss.pop,"_minMAC2_summary_pij.out_pop.info"), select=c("Population")))$Population
  pops=pops[order(pops)]
  ### Coverage
  coverage[[subsp]]=fread(paste0(allele_freq_dir, "/f5", 
                                 subsp.file, ".pops_minInd.6or50pct_missing.pops.", miss.pop,"/f5", subsp.file, 
                                 ".pops.all.chrs_missing.pops.", miss.pop,"_minMAC2_coverage"))
  ### Population colours (according to subspecies)
  subsp.col[[subsp]]=c()
  for(pop in pops){
    if(startsWith(pop, "c.")){subsp.col[[subsp]]=c(subsp.col[[subsp]], "green3")}
    if(startsWith(pop, "e.")){subsp.col[[subsp]]=c(subsp.col[[subsp]], "darkorange")}
    if(startsWith(pop, "n.")){subsp.col[[subsp]]=c(subsp.col[[subsp]], "red")}
    if(startsWith(pop, "w.")){subsp.col[[subsp]]=c(subsp.col[[subsp]], "blue")}
  }
  ## Read BayPass AUX output
  ### Annotated Betai
  #betai.ann[[subsp]]=fread(paste0(formatted_aux_out_dir, "/f5",
  #                       subsp.file ,".pops_missing.pops.", miss.pop,"_minMAC2_", env_data, "_summary_betai.out_row.per.gtf.annot_5000bp.flanks_all_runs.gz"))
  #betai.ann[[subsp]][betai.ann[[subsp]]$gene=='.','gene']=NA
  ### Not annotated Betai
  betai[[subsp]]=fread(paste0(formatted_aux_out_dir, "/f5",
                         subsp.file ,".pops_missing.pops.", miss.pop,"_minMAC2_", env_data, "_summary_betai.out_all_runs.gz"))
  ### Pdelta file
  Pdelta[[subsp]]=fread(paste0(formatted_aux_out_dir, "/f5",
                         subsp.file ,".pops_missing.pops.", miss.pop,"_minMAC2_", env_data, "_summary_Pdelta.out_all_runs"))
  ## Read in covariable input files - real values and random ones
  ### Get the number of random reps from column headers
  random_reps=colnames(betai[[subsp]])[grepl("RANDOM", colnames(betai[[subsp]]))]
  random_reps=as.numeric(unique(gsub(".*RANDOM_rep", "", random_reps)))
  for(rep in c('Real', random_reps)){
    if(rep=='Real'){
      rep.name=rep
      cov.in[[subsp]][[rep.name]]=fread(paste0("../../running_baypass/baypass_env_input/", env_data, ".", subsp, ".baypass_env_input.txt"))
    }
    if(rep %in% random_reps){
      rep.name=paste0("Random_", rep)
      cov.in[[subsp]][[rep.name]]=fread(paste0("../../running_baypass/baypass_env_input/", 
                                                  env_data ,".", subsp, ".baypass_env_input.RANDOM_rep", rep,".txt"))
    }
    ### Add covariable names as row names (alphabetical order)
    covs=unique(Pdelta[[subsp]]$COVARIABLE_name)
    names(habitat_col)=covs[order(covs)]
    rownames(cov.in[[subsp]][[rep.name]])=covs[order(covs)]
    ### Add population names as column names (alphabetical order)
    colnames(cov.in[[subsp]][[rep.name]])=pops
  }
}
```

## Covariable Map

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
if('all' %in% subsps){
  subsps2='all'
}else{
  subsps2=subsps
}
for(subsp in subsps2){
  env_var_map(merge(env_file.pops, data.frame(t(cov.in[[subsp]][['Real']])), by.x="Population", by.y=0, all.y=TRUE), col=col_scale)
}
```

# Review output files

Information comes from the 2019 BayPass manual.

BayPass output files we formatted with `/Users/harrisonostridge/OneDrive - University College London/Projects/PanAf/phase1and2_exomes/baypass/baypass_aux/scripts/format_outputformat_baypass_aux_output.Rmd`.

## summary_Pdelta.out file

M_P and SD_P: "posterior mean and the standard deviation of the parameter P corresponding to the overall proportion of SNPs associated with each given covariable."

- aP and bP define the prior odds on P. M_P and SD_P correspond to the posterior estimates of P

I think M_P is very informative as it tells us how much evidence there is for association to a habitat across the genome.

```{r echo=FALSE, out.width="25%"}
plot_Pdelta.subsp_panels(Pdelta)
```

## Summary_betai.out file

M_Beta and SD_Beta: posterior mean and the standard deviation of the regression coefficient β.

PIP: posterior mean of the auxiliary variable.

BF(dB): estimated Bayes Factor in dB units (10×log10(BF)).

Here I plot the results from a single run.

```{r echo=FALSE, out.width="25%"}
plot_betai_stats(betai)
```

# Plot Median Values

From here on, I plot median values over three independent runs with different seeds to account for run-to-run variation.

## Volcano Plot

```{r echo=FALSE}
min.M_P=0.005
```

NB: if there are ~0 SNPs associated with a habitat (i.e. the posterior mean of the parameter P is < `r paste0(min.M_P)`), I do not include it in the graph as it messes up the axis scales.

#### BF

NB: An apparent excess of SNPs where BF=53 may occur because this is the maximum BF.

  - I think this is because if the posterior mean of the axillary variable (delta) is 1 or 0, then to account for the finite number T of MCMC sampled values, the posterior mean of delta is set to (T-0.5)/(T-1) or 0.5/(T-1) respectively. I think this is to account for the fact that we cannot be 100% certain an association is true (or not) as we haven't sampled infinite MCMC runs?
  
BF distributions tend to reflect M_P estimates with higher M_P values corresponding to BF distributions shifted to higher values or with longer tails.

#### Beta

NB: The original BayPass paper describes beta > 0.2 as very strong associations and betas < 0.1 as weak associations.

NB: The density distribution of beta values is shown on a log scale (to account for the vast excess of sites with beta ≈ 0) and so differences between habitat distributions are more pronounced than they might first appear.

### Plot 

```{r echo=FALSE, fig.align='center'}
for(subsp in subsps){
  cat(subsp, "\n")
  # Select only habitats with any evidence of selection (otherwise it makes the graph hard to read)
  ## Select only results from real (not random) runs
  Pdelta_plot_df=Pdelta[[subsp]][Pdelta[[subsp]]$Run=='real',]
  ## Get mean across subsets
  Pdelta_plot_df=aggregate(Pdelta_plot_df['M_P.median'], Pdelta_plot_df[c("COVARIABLE_name")], mean)
  covs=Pdelta_plot_df[Pdelta_plot_df$M_P.median>min.M_P,]$COVARIABLE_name
  # Plot volcano plot
  aux_volcano.v2(betai[[subsp]][betai[[subsp]]$COVARIABLE_name %in% covs,], x='M_Beta.median', y='`BF(dB).median`', colour_col='COVARIABLE_name', colours=habitat_col[covs])
}
```

- there is a clear skew towards negative beta values in all samples and central-eastern.

## Distribution Across the Genome

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.asp=0.3, out.width="100%"}
for(subsp in subsps){
  # All sites and chrs
  manhattan_plot(betai[[subsp]], cov_col=habitat_col, stat="`BF(dB).median`", title=paste0(subsp, " Manhattan Plot"))
}
```

## Correlation with XtX

```{r echo=FALSE, message=FALSE, warning=FALSE, out.width="33%"}
core_vs_aux(betai)
```


