#### Setup

```{r setup}
rm(list=ls())
knitr::opts_knit$set(root.dir = '~/OneDrive - University College London/Projects/Ostridge_PanAf/baypass/analysing_baypass_output/baypass_aux/') 
```

#### Parameters

```{r}
subsps=c('all','ce', 'w')
subsps_title=c('all'='All Subspecies','ce'='Central-Eastern','n'='Nigeria-Cameroon', 'w'='Western')
env_data='f_over_sum_known_trees'
habitat_col='turquoise4'
tails=c(0.005, 0.001, 0.0005)
chr21=TRUE
flanks=1000
only_unique_snps=FALSE
```

```{r echo=FALSE}
if(chr21){
  title_null="chr21 null"
  suffix=paste0(".non-genic_",flanks,".flanks")
}else{
  title_null="random shuffling null"
  suffix="-cov_cor"
}
```

---
title: "`r paste0("BayPass AUX Distribution of Candidates in the Genome - ", env_data, ", ", title_null)`"
subtitle: "`r paste0("Subspecies datasets: ", paste(subsps, collapse = ", "))`"
author: "Harrison Ostridge"
date: "`r Sys.Date()`"
output: html_document
---

This script looks at the distribution of physical positions of candidate SNPs in the genome. We are mostly looking for evidence 'peaks' resulting on linkage disequilibrium with SNPs under selection.

Candidate SNPs were selected in `r paste0("scripts/", env_data, "/", env_data,".baypass_aux_candidate.Rmd")` and saved to a file (`r `).

#### Library

```{r library, message=FALSE, warning=FALSE}
library(data.table)
options(datatable.fread.datatable=FALSE)
library(dplyr)
source("../scripts/baypass_tools.R")
source("scripts/baypass_aux_tools.R")
source("../baypass_core/scripts/distribution_of_candidates_in_the_genome_tools-cov_cor.R")

env_dir="../../../environmental_data/output/"
formatted_aux_out_dir="output/formatted_baypass_aux_output/"
formatted_aux_out_fprs_dir="output/baypass_aux_output_with_fprs/"
annotation_file="../../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs.gtf"
```

#### Read in data

```{r echo=FALSE}
## Environmental data file
env_file=read.csv(paste0(env_dir, "meta_data.env.pops.imputed.csv"))

# Subspecies specific files
betai_fpr=list()
betai_fpr.ann=list()

if(chr21){
  suffix=paste0(".non-genic_",flanks,"bp.flanks")
}else{
  suffix="-cov_cor"
}

for(subsp in subsps){
  cat("-", subsp, "\n")
  if(subsp=='all'){subsp.file=''}else{subsp.file=paste0(".", subsp)}
  if(subsp=='n'){miss.pop=0}else{miss.pop=0.3}
  ## Read BayPass AUX output
  ### Annotated Betai
  ann=unique(fread(paste0(formatted_aux_out_dir, "/f5",
                         subsp.file ,".pops_missing.pops.", miss.pop,"_minMAC2_", env_data, "_summary_betai.out_row.per.gtf.annot_5000bp.flanks_all_runs.gz"),
                         select=c('chr', 'pos', 'gene')))
  ann[ann$gene=='.','gene']=NA
  ### FPR data
  betai_fpr[[subsp]]=unique(fread(paste0(formatted_aux_out_fprs_dir, "/f5", subsp.file ,".pops_missing.pops.", miss.pop,"_minMAC2_", 
                                         env_data, "_summary_betai.out_all_runs.fpr", suffix)))
  betai_fpr.ann[[subsp]]=merge(betai_fpr[[subsp]], ann, by=c('chr', 'pos'), all.x=TRUE)
  betai_fpr.ann[[subsp]]$chr_pos=paste(betai_fpr.ann[[subsp]]$chr, betai_fpr.ann[[subsp]]$pos, sep="_")
  # Do not allow SNPs to belong to multiple genes, just select the one that comes first alphebetically
  if(only_unique_snps){
    betai_fpr.ann[[subsp]]$chr_pos=paste(betai_fpr.ann[[subsp]]$chr, betai_fpr.ann[[subsp]]$pos, sep="_")
    betai_fpr.ann[[subsp]]=betai_fpr.ann[[subsp]][order(betai_fpr.ann[[subsp]]$gene),]
    betai_fpr.ann[[subsp]]=betai_fpr.ann[[subsp]][!duplicated(betai_fpr.ann[[subsp]]$chr_pos),]
  }
  ann=NULL
}
gtf=read.table(annotation_file)
covs=unique(betai_fpr[[1]]$COVARIABLE_name)
```

# Manhatten plot

```{r fig.height=3, fig.width=8}
source("../scripts/baypass_tools.R")
for(subsp in subsps){
  #cat(subsp, "\n")
  # Prepare lists of datasets with candidates (to highlight in plot)
  focal.snps.list=list()
  for(tail in tails){
    focal.snps.list=c(focal.snps.list, list(betai_fpr[[subsp]][betai_fpr[[subsp]]$fpr<tail,]))
  }
  # Add new columns
  colnames(betai_fpr[[subsp]])[colnames(betai_fpr[[subsp]])=='BF(dB).median']="BF.med"
  betai_fpr[[subsp]]$log_fpr=-log(betai_fpr[[subsp]]$fpr)
  # Plot
  for(stat in c('log_fpr', 'BF.med')[2]){
    manhattan_plot(betai_fpr[[subsp]], stat=stat, title=paste0(subsps_title[[subsp]]), cov_col=habitat_col,
                 focal.snps.list = focal.snps.list)
  }
}
```

# SNPs per gene

```{r out.width="50%"}
snps_per_gene=list()
genes_per_snp=list()
n_snps_genes=data.frame("Dataset"=character(), "Facet"=numeric(), "Tail_pct"=numeric(), "N SNPs"=numeric(), "N genes"=numeric())
for(subsp in subsps){
  for(tail in c(1, seq(0.01, 0.0000, by=-0.0001))){
    if(tail==1){facet="Total"}else{facet="Tails"}
    cands=betai_fpr.ann[[subsp]][betai_fpr.ann[[subsp]]$fpr<tail,]
    n_snps_genes=rbind(n_snps_genes, data.frame("Dataset"=subsp, "Tail_pct"=100*tail, "Facet"=facet, 
                                                "N_SNPs"=nrow(unique(cands[,c('chr', 'pos')])), 
                                                "N_genes"=length(unique(cands$gene))))
  }
  snps_per_gene[[subsp]]=data.frame("gene"=character(), "N_SNPS"=numeric(), "Tail"=character())
  genes_per_snp[[subsp]]=data.frame("gene"=character(), "N_genes"=numeric(), "Tail"=character())
  for(tail in tails){
    cands=betai_fpr.ann[[subsp]][betai_fpr.ann[[subsp]]$fpr<tail,]
    cands$chr_pos=paste(cands$chr, cands$pos, sep="_")
    
    snps_per_gene.tmp=data.frame(table(cands$gene))
    colnames(snps_per_gene.tmp)=c('gene', 'N_SNPS')
    snps_per_gene.tmp$Tail=paste0(100*tail, "%")
    snps_per_gene[[subsp]]=rbind(snps_per_gene[[subsp]], snps_per_gene.tmp)
    
    genes_per_snp.tmp=data.frame(table(cands$chr_pos))
    colnames(genes_per_snp.tmp)=c('gene', 'N_genes')
    genes_per_snp.tmp$Tail=paste0(100*tail, "%")
    genes_per_snp[[subsp]]=rbind(genes_per_snp[[subsp]], genes_per_snp.tmp)
  }
  # Plot
  snps_per_gene_hist=ggplot(snps_per_gene[[subsp]], aes(x=N_SNPS, y=..density.., col=Tail, fill=Tail)) +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5)) +
    labs(title=paste0(subsps_title[[subsp]]), y="Density", x="Number of SNPs per gene") +
    stat_bin(alpha=1, binwidth=1, geom="step", position="identity") +
    scale_color_manual(values = c("darkorange", "red", "purple"))
  print(snps_per_gene_hist)
  
  genes_per_snp_hist=ggplot(genes_per_snp[[subsp]], aes(x=N_genes, y=..density.., col=Tail, fill=Tail)) +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5)) +
    labs(title=paste0(subsps_title[[subsp]]), y="Density", x="Number of genes per SNP") +
    stat_bin(alpha=1, binwidth=1, geom="step", position="identity") +
    scale_color_manual(values = c("darkorange", "red", "purple"))
  print(genes_per_snp_hist)
}
```

There seems generally to be a skew to fewer genes per SNP and fewer SNPs per gene at more stringent thresholds. There is presumably a greater skew towards fewer SNPs per gene which explains the reduction in the SNPs-gene ratio as you move to more stringent tails.

```{r fig.height=3, fig.width=10}
null=TRUE
reps=50
span=0.25
alpha=0.25

n_snps_genes$SNPs_per_gene=n_snps_genes$N_SNPs/n_snps_genes$N_gene

n_snps_genes[n_snps_genes$Dataset=='all', "Dataset"]="All subspecies"
n_snps_genes[n_snps_genes$Dataset=='ce', "Dataset"]="Central-eastern"
n_snps_genes[n_snps_genes$Dataset=='n', "Dataset"]="Nigeria-Cameroon"
n_snps_genes[n_snps_genes$Dataset=='w', "Dataset"]="Western"

snps_per_gene_plot=ggplot(n_snps_genes[n_snps_genes$Tail_pct!=100,], aes(y=SNPs_per_gene, x=Tail_pct, col=Dataset, fill=Dataset)) +
  theme_linedraw() +
      theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5, size=15),
            panel.grid.minor.x = element_blank(),panel.grid.minor.y = element_blank(),
            strip.background =element_rect(fill="white"), strip.text = element_text(colour = 'black'),) +
  #facet_grid(~Facet, scales = "free",space="free_y") +
  labs(title="SNPs per Gene", x="FPR tail (%)", y="SNPs per Gene") +
  #geom_point() +
  #geom_line(aes(linetype="Data"))
  geom_smooth(aes(linetype="Data"), method = 'loess', se=T, span = span, alpha=alpha)

if(null){
  null_df=data.frame("Dataset"=character(), "Tail_pct"=numeric(), "rep"=numeric(), "SNPs_per_gene"=numeric())
  for(subsp in subsps){
    if(subsp=='all'){dataset="All subspecies"}
    if(subsp=='ce'){dataset="Central-eastern"}
    if(subsp=='n'){dataset="Nigeria-Cameroon"}
    if(subsp=='w'){dataset="Western"}
    for(tail in unique(n_snps_genes$Tail_pct)[unique(n_snps_genes$Tail_pct)!=100]){
      n_snps_genes.tmp=n_snps_genes[n_snps_genes$Dataset==dataset & n_snps_genes$Tail_pct==tail, ]
      chr_pos.subsp=unique(betai_fpr.ann[[subsp]]$chr_pos)
      for(rep in 1:reps){
        set.seed(rep)
        chr_pos.tmp=sample(chr_pos.subsp, n_snps_genes.tmp$N_SNPs, replace=FALSE)
        n_genes=length(unique(betai_fpr.ann[[subsp]][betai_fpr.ann[[subsp]]$chr_pos %in% chr_pos.tmp, ]$gene))
        null_df=rbind(null_df, data.frame("Dataset"=dataset, "Tail_pct"=tail, "rep"=rep, "SNPs_per_gene"=n_snps_genes.tmp$N_SNPs/n_genes))
        set.seed(NULL)
      }
    }
  }
  snps_per_gene_plot=snps_per_gene_plot+
    geom_smooth(data=null_df, aes(y=SNPs_per_gene, x=Tail_pct, col=Dataset, fill=Dataset, linetype="Null"), method = 'loess', se=T, span = span, alpha=alpha) +
    scale_discrete_manual(name="", aesthetics=c('linetype'), values = c("Data"='solid', "Null"='dotted'))
  
}

snps_per_gene_plot=snps_per_gene_plot +
  scale_x_continuous(breaks=c(100, 100*seq(0.01, 0.0005, by=-0.001))) +
  scale_discrete_manual(aesthetics=c('colour', 'fill'), values = c("All subspecies"='black', "Central-eastern"='brown4', "Nigeria-Cameroon"='red', "Western"='blue')) +
  guides(linetype=guide_legend(override.aes=list(fill=NA))) +
  #scale_x_log10()
  scale_x_continuous(limits = c(0,1), breaks=c(seq(0, 1, 0.1)))


snps_per_gene_grob=ggplotGrob(snps_per_gene_plot)
snps_per_gene_grob$widths[5] = unit(8.5, "cm")
snps_per_gene_grob$widths[9] = unit(3, "cm")

# Plot the graph
grid.newpage()
grid.draw(snps_per_gene_grob)
```
