---
title: "Filter non-genic SNPs"
author: "Harrison Ostridge"
date: "`r Sys.Date()`"
output: html_document
---

The purpose of this script is to annotate SNPs from chr21 and then select those which are non-genic. We want to do this to generate 'neutral' genetic data to compare to the exomes.

#### Setup

```{r setup}
rm(list=ls())
knitr::opts_knit$set(root.dir = '~/OneDrive - University College London/Projects/Ostridge_PanAf/allele_frequencies/chr21/') 
```


```{r library, message=FALSE, warning=FALSE}
library(data.table)
options(datatable.fread.datatable=FALSE)
library(dplyr)
library(tidyverse)

#source("../../phase1and2_exomes/baypass/baypass_core/baypass_core_inference/scripts/plotting_tools/distribution_of_candidates_in_the_genome_tools-cov_cor.R")
```

# scp results from myriad

```{bash eval=FALSE}
for DATASET in chr21.f7.c.pops chr21.f7.e.pops 
  do
  for MISS_POPS in 0 0.3
    do
    # I have to add this just because of a stupid naming convention
    if [[ $MISS_POPS == 0 ]]
        then
                MISS_POPS2=0.0
        else
                MISS_POPS2=$MISS_POPS
        fi
    # scp
    scp ucfajos@myriad.rc.ucl.ac.uk:/home/ucfajos/Scratch/output/phase1and2_chr21_output/allele.frequencies/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_\*_minMAC2 output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/
  done
done
```

# Annotate SNPs

### GTF file

The GTF file was downloaded from http://ftp.ensembl.org/pub/grch37/current/gtf/homo_sapiens/Homo_sapiens.GRCh37.87.chr.gtf.gz. The gene_id column was replaced with the gene_name column (which contained gene symbols) in `/Users/harrisonostridge/OneDrive - University College London/Projects/PanAf/phase1and2_exomes/gowinda/scripts/run_gowinda.v11.Rmd`. 

## BEDtools intersect

-v means "Only report those entries in A that have no overlap in B"

BEDtools manual (useful): https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html 

I first used a flank of 5000 as that is what we tend to do to define a SNP as belonging to a gene. However, in order to increase the number of SNPs in the null (and allow us to move further into the empirical tail when using the null) we needed more SNPs so I tried again with 1000bp flanks. This will only make the null slightly more stringent.

- Using 1kbp rather than 5kbp around genes results in only 3% more non-genic SNPs so the gains are marginal.

```{bash}
FLANK=1000

# Of the GTP file with falnks added does not exist
if ! test -f "../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs_${FLANK}bp.flanks.gtf"; then
  # Add flanks to GTF file
  awk -v OFS='\t' -v flank="$FLANK" '{print $1,$2,$3,$4-flank,$5+flank,$6,$7,$8,$9,$10}'  ../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs.gtf \
  > ../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs_${FLANK}bp.flanks.gtf
fi

# BEDtools intersect
for DATASET in chr21.f7.c.pops chr21.f7.e.pops 
  do
  echo '### '${DATASET}
  for MISS_POPS in 0 0.3
    do
    echo '#### Missing populations: '${MISS_POPS}
    # I have to add this just because of a stupid naming convention
    if [[ $MISS_POPS == 0 ]]
        then
                MISS_POPS2=0.0
        else
                MISS_POPS2=$MISS_POPS
        fi
    # Make allele counts file into bed format (have an 'end' column remove column headers and remove 'chr' prefix)
    ## Also select only sites which map to chr21 (some reads map to other chromosomes)
    awk '$1 == "chr21"' output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2 | \
    awk -F '\t' -v OFS='\t' '{print $1, $2, $2+1, substr($0, index($0,$3))}' | \
    sed '1d' | \
    sed 's/chr//g' > \
    output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.bed
  
    # Bedtools intersect
    ## Extract column headers (modified bedtools output will be appended onto this)
    head -n 1 output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2 > \
    output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.non-genic_${FLANK}bp.flanks
    ## Run bedtools
    bedtools intersect \
    -a output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.bed \
    -b ../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs_${FLANK}bp.flanks.gtf \
    -v | \
    awk -F '\t' -v OFS='\t' '{print $1, $2, substr($0, index($0,$4))}' >> \
    output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.non-genic_${FLANK}bp.flanks
  
    # Make BayPass input file (remove chr, start and end columns, remove headers, and ensure it is space delineated (and remove leading whitespace introduced with `awk -v OFS=' ' '{$1=$2=""; print $0}'` using sed 's/^ *//g'))
    awk -v OFS=' ' '{$1=$2=""; print $0}' \
    output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.non-genic_${FLANK}bp.flanks |
    sed 's/^ *//g' | \
    sed '1d' > \
    output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_baypass.input.geno_minMAC2.non-genic_${FLANK}bp.flanks
  
    echo "Total N SNPs: " `cat output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_baypass.input.geno_minMAC2 | wc -l`
    echo "Non-genic N SNPs: " `cat output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_baypass.input.geno_minMAC2.non-genic_${FLANK}bp.flanks | wc -l`
    head -n 3 output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_pop.allele.counts_minMAC2.non-genic_${FLANK}bp.flanks

    echo ''
  done
done
```

Slightly more than half the SNPs are within 5kb of a gene.

# Check it worked

```{r message=FALSE, warning=FALSE, out.width="33%"}
gtf=fread('../../gowinda/baypass_core/output/gowinda_input/annotation_files/Homo_sapiens.GRCh37.87.chr_gene.symbol_sync_homologs_5000bp.flanks.gtf', sep="\t")
example=fread('output/chr21.f7.w.pops_minInd.6or50pct_missing.pops.0.3/chr21.f7.w.pops_chr21_missing.pops.0.3_pop.allele.counts_minMAC2.non-genic_5000bp.flanks', select=c('chr', 'pos'))

plots=list()
plots[[1]]=SNP_distribution.v2(example[order(example$pos),], gtf=gtf)
plots[[2]]=SNP_distribution.v2(tail(example[order(example$pos),], 10000), gtf=gtf)
plots[[3]]=SNP_distribution.v2(tail(example[order(example$pos),], 1000), gtf=gtf)
for(plot in plots){print(plot)}
```

Red represents genes, we can see all sites are in non-genic regions.

# scp to myriad

```{bash eval=FALSE}
for DATASET in chr21.f7.c.pops chr21.f7.e.pops 
  do
  for MISS_POPS in 0 0.3
    do
    # I have to add this just because of a stupid naming convention
    if [[ $MISS_POPS == 0 ]]
        then
                MISS_POPS2=0.0
        else
                MISS_POPS2=$MISS_POPS
        fi
    rsync --progress --human-readable output/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/${DATASET}_chr21_missing.pops.${MISS_POPS2}_*_minMAC2.non-genic_*.flanks ucfajos@myriad.rc.ucl.ac.uk:/home/ucfajos/Scratch/output/phase1and2_chr21_output/allele.frequencies/${DATASET}_minInd.6or50pct_missing.pops.${MISS_POPS}/
  done
done
```

